### PR TITLE
Improve syntax highlighting by adding a tree-sitter parser

### PR DESCRIPTION
**Pull request by nedpals**
Adds better syntax highlighting capabilities using the [tree-sitter parser](http://tree-sitter.github.io/tree-sitter/) (as seen on the screenshot below).

image

For now, it uses the code from the [vscode-tree-sitter](https://github.com/georgewfraser/vscode-tree-sitter) extension to activate its highlighting capabilities since VS Code has not implemented yet any option for enabling custom tokenization and colorization engines aside from TextMate's.

A separate directory inside the `src` folder named `tree-sitter` has been created instead of a single file (like `highlight.v`) to have a single folder for implementing features where the [tree-sitter parser for V](https://github.com/nedpals/tree-sitter-v) would be badly needed aside from syntax highlighting.

Aside from that, it would be a huge improvement over the maintainance of syntax highlighting. For now, the initial code is a bit ugly and taken from Go. Hence, edge cases in syntax are not covered yet.

**NOTE:** The TextMate grammar has been kept as a fallback.

### CODE REVIEW

**Comment by Alhadis** [comment]
> Aside from that, it would be a huge improvement over the maintainance of syntax highlighting. For now, the initial code is a bit ugly and taken from Go. Hence, edge cases in syntax are not covered yet.

I've written over 30+ grammars by hand in CSON and never had a problem with maintenance, ever. I'd happily rewrite this package's grammar if @0x9ef wanted me to.

IMHO, tree-sitter is horribly over-engineered and introduces a shitload of complexity that shouldn't have to exist, especially for something as simple as syntax highlighting.

File: package.json [hunk 1]
```diff
@@ -152,11 +152,14 @@
 	],
 	"main": "./out/extension.js",
 	"dependencies": {
-		"vscode-languageclient": "~5.2.1"
+		"jsonc-parser": "^2.2.0",
+		"tar": "^5.0.5",
```
Comment (alexesprit) [inline package.json]: Where is `tar` used?

Comment (nedpals) [reply package.json]: Just read the source code again and I was never aware the usage of tar even in the original code. I removed it. Thanks.

**Comment by nedpals** [comment]
> IMHO, tree-sitter is horribly over-engineered and introduces a shitload of complexity that shouldn't have to exist, especially for something as simple as syntax highlighting.

Care to elaborate on this? AFAIK the regex hell TextMate grammar suffers (hence, the maintenance I'm referring to) are one of the reasons Tree-Sitter are preferred.

**Comment by spytheman** [comment]
@Alhadis what is CSON in this context?

**Comment by Alhadis** [comment]
@spytheman CSON is [CoffeeScript Object Notation](https://www.npmjs.com/package/cson), a format that compiles to JSON and is used by Atom packages for configuration and authoring grammars. [Here are some examples](https://github.com/Alhadis/language-etc/tree/master/grammars).

**Comment by 0x9ef** [comment]
Hello! How to test it?

**Comment by nedpals** [comment]
@0x9ef Hi! You can try it by pressing `F5`, the TS output will show up, then create a new file (don't save it), paste a random code, then set the language mode to `V`. Like this:
image

**Comment by dhonx** [comment]
I have the same result @nedpals

**Comment by alexesprit** [comment]
It works for me (tokens in a red square are highlighted now):
image

But it works incorectly for operator overloads:
image
1. The return type should be highlighted in a different color (for my color theme).
2. The type of a left operand is not highlighted.

**Comment by spytheman** [comment]
@Alhadis thank you

**Comment by dhonx** [comment]
Can we close this @nedpals ?

**Comment by nedpals** [comment]
sure i have moved on already lol

**Comment by danieldaeschle** [comment]
@Alhadis Are you still interested in improving the syntax highlighting?

**Comment by Alhadis** [comment]
@danieldaeschle Yes, sorry. Just haven't gotten around to it yet.

I'll see what I can do over the holidays. 👍

**Comment by danieldaeschle** [comment]
Thanks :)

**Comment by Alhadis** [comment]
Progress can be tracked [here](https://github.com/Cutlery-Drawer/language-v/tree/rewrite). Bear in mind I only just started last night. 😉

When I'm done, I'll submit a PR to this repository along with logic to recompile the CSON grammar to a JSON file, perhaps automatically when saved. I don't use VSCode, so I'm unfamiliar with whatever APIs would be used to achieve that. 😕

**Comment by nedpals** [comment]
> When I'm done, I'll submit a PR to this repository along with logic to recompile the CSON grammar to a JSON file, perhaps automatically when saved. I don't use VSCode, so I'm unfamiliar with whatever APIs would be used to achieve that.

@Alhadis No worries, we will take care of that :slightly_smiling_face:

**Pull request closed without merging**